File: app/src/main/java/ninja/dudley/yamr/ui/fragments/ProviderViewer.java
package ninja.dudley.yamr.ui.fragments;

import android.app.Activity;
import android.app.ListFragment;
import android.app.LoaderManager;
import android.app.ProgressDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.CursorLoader;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.Loader;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.content.LocalBroadcastManager;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.SearchView;
import android.widget.SimpleCursorAdapter;

import ninja.dudley.yamr.R;
import ninja.dudley.yamr.coms.LoadSeries;
import ninja.dudley.yamr.db.DBHelper;
import ninja.dudley.yamr.model.Provider;
import ninja.dudley.yamr.model.Series;
import ninja.dudley.yamr.svc.FetcherAsync;

public class ProviderViewer extends ListFragment
        implements LoaderManager.LoaderCallbacks<Cursor>, SearchView.OnQueryTextListener
{
    private SimpleCursorAdapter adapter;

    private BroadcastReceiver fetchStatusReceiver;
    private BroadcastReceiver fetchCompleteReceiver;

    private LoadSeries parent;

    private ProgressDialog loading;

    private String filter;

    @Override
    public void onAttach(Activity activity)
    {
        super.onAttach(activity);

        this.parent = (LoadSeries) activity;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
    {
        setHasOptionsMenu(true);

        LinearLayout layout = (LinearLayout) inflater.inflate(R.layout.fragment_provider_viewer, container, false);

        fetchStatusReceiver = new BroadcastReceiver()
        {
            @Override
            public void onReceive(Context context, Intent intent)
            {
                float percent = 100 * intent.getFloatExtra(FetcherAsync.FETCH_PROVIDER_STATUS, 0.0f);
                loading.setProgress((int) percent);
            }
        };

        fetchCompleteReceiver = new BroadcastReceiver()
        {
            @Override
            public void onReceive(Context context, Intent intent)
            {
                getLoaderManager().restartLoader(0, null, ProviderViewer.this);
                adapter.notifyDataSetChanged();
                loading.dismiss();
            }
        };

        Intent i = new Intent(getActivity(), FetcherAsync.class);
        i.setAction(FetcherAsync.FETCH_PROVIDER);
        i.setData(Provider.uri(1));   // Hard-code to the first (mangapanda) for now.
        getActivity().startService(i);

        adapter = new SimpleCursorAdapter(
                getActivity(),
                R.layout.series_item,
                null,
                new String[]{DBHelper.SeriesEntry.COLUMN_NAME},
                new int[]{R.id.series_name},
                0
        );
        setListAdapter(adapter);

        getLoaderManager().initLoader(0, null, this);
        loading = new ProgressDialog(getActivity());
        loading.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
        loading.setTitle("Loading Series");
        loading.show();
        return layout;
    }

    @Override
    public void onResume()
    {
        super.onResume();
        IntentFilter completeFilter = new IntentFilter(FetcherAsync.FETCH_PROVIDER_COMPLETE);
        try
        {
            completeFilter.addDataType(getActivity().getContentResolver().getType(Provider.baseUri()));
        }
        catch (IntentFilter.MalformedMimeTypeException e)
        {
            // I'm a little more OK with this, as Provider.baseUri() is static.
            throw new AssertionError(e);
        }
        LocalBroadcastManager.getInstance(getActivity()).registerReceiver(fetchCompleteReceiver, completeFilter);
        LocalBroadcastManager.getInstance(getActivity()).registerReceiver(fetchStatusReceiver, new IntentFilter(FetcherAsync.FETCH_PROVIDER_STATUS));
    }

    @Override
    public void onPause()
    {
        super.onPause();
        LocalBroadcastManager.getInstance(getActivity()).unregisterReceiver(fetchStatusReceiver);
        LocalBroadcastManager.getInstance(getActivity()).unregisterReceiver(fetchCompleteReceiver);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater)
    {
        inflater.inflate(R.menu.menu_provider_viewer, menu);
        MenuItem item = menu.findItem(R.id.search);
        SearchView sv = (SearchView) item.getActionView();
        sv.setOnQueryTextListener(this);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        switch (item.getItemId())
        {
            case R.id.search:
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public Loader<Cursor> onCreateLoader(int id, Bundle args)
    {
        Uri mangaPandaSeries = Provider.uri(1).buildUpon().appendPath("series").build();
        if (filter == null)
        {
            return new CursorLoader(
                    getActivity(),
                    mangaPandaSeries,
                    null,
                    null,
                    null,
                    null
            );
        }
        else
        {
            return new CursorLoader(
                    getActivity(),
                    mangaPandaSeries,
                    null,
                    DBHelper.SeriesEntry.COLUMN_NAME + " like ?",
                    new String[]{'%' + filter + '%'},
                    null
            );
        }
    }

    @Override
    public void onLoadFinished(Loader<Cursor> loader, Cursor data)
    {
        adapter.changeCursor(data);
    }

    @Override
    public void onLoaderReset(Loader<Cursor> loader)
    {
        adapter.changeCursor(null);
    }

    @Override
    public void onListItemClick(ListView l, View v, int position, long id)
    {
        super.onListItemClick(l, v, position, id);
        parent.loadSeries(Series.uri((int) id));
    }

    @Override
    public boolean onQueryTextSubmit(String query)
    {
        return true;
    }

    @Override
    public boolean onQueryTextChange(String newText)
    {
        filter = !TextUtils.isEmpty(newText) ? newText : null;
        getLoaderManager().restartLoader(0, null, this);
        return true;
    }
}

File: app/src/main/kotlin/ninja/dudley/yamr/ui/fragments/ProviderViewer.kt
package ninja.dudley.yamr.ui.fragments

import android.app.Activity
import android.app.ListFragment
import android.app.LoaderManager
import android.app.ProgressDialog
import android.content.Context
import android.content.CursorLoader
import android.content.Loader
import android.database.Cursor
import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import android.view.*
import android.view.inputmethod.InputMethodManager
import android.widget.LinearLayout
import android.widget.ListView
import android.widget.SearchView
import android.widget.SimpleCursorAdapter
import ninja.dudley.yamr.R
import ninja.dudley.yamr.model.Provider
import ninja.dudley.yamr.model.Series
import ninja.dudley.yamr.svc.FetcherAsync
import ninja.dudley.yamr.svc.FetcherSync
import ninja.dudley.yamr.ui.activities.Browse

// Method reference functions
public fun providerViewerStatus(providerViewer: Any, status: Float)
{
    (providerViewer as ProviderViewer).status(status)
}

public fun providerViewerComplete(providerViewer: Any, provider: Provider)
{
    (providerViewer as ProviderViewer).complete(provider)
}

public class ProviderViewer :
        ListFragment(), LoaderManager.LoaderCallbacks<Cursor>, SearchView.OnQueryTextListener
{
    private var adapter: SimpleCursorAdapter? = null

    private var loading: ProgressDialog? = null
    private var filter: String? = null

    private var parent: Browse? = null

    private var providerUri: Uri? = null

    override fun onAttach(activity: Activity?)
    {
        super.onAttach(activity)
        this.parent = activity as Browse?
    }

    fun status(status: Float)
    {
        if (!isAdded)
        {
            return
        }
        val percent = 100 * status
        loading!!.progress = percent.toInt()
    }

    fun complete(provider: Provider)
    {
        if (!isAdded)
        {
            return
        }
        this.providerUri = provider.uri()
        loaderManager.restartLoader(0, Bundle(), this@ProviderViewer)
        adapter!!.notifyDataSetChanged()
        loading!!.dismiss()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View?
    {
        setHasOptionsMenu(true)

        val layout = inflater.inflate(R.layout.fragment_provider_viewer, container, false) as LinearLayout

        this.providerUri = Uri.parse(arguments.getString(PROVIDER_ARG_KEY))
        val provider = Provider(activity.contentResolver.query(providerUri, null, null, null, null))
        FetcherAsync.fetchProvider(provider, this, ::providerViewerComplete, ::providerViewerStatus)

        adapter = SimpleCursorAdapter(activity,
                                     R.layout.simple_series_item,
                                     null,
                                     arrayOf(Series.nameCol),
                                     intArrayOf(R.id.series_name), 0)
        listAdapter = adapter

        loaderManager.initLoader(0, Bundle(), this)
        loading = ProgressDialog(activity)
        loading!!.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL)
        loading!!.setTitle("Loading Series")
        loading!!.show()
        return layout
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?)
    {
        inflater!!.inflate(R.menu.menu_provider_viewer, menu)
        val item = menu!!.findItem(R.id.search)
        val sv = item.actionView as SearchView
        sv.setOnQueryTextListener(this)
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean
    {
        when (item!!.itemId)
        {
            R.id.search -> return true
            R.id.refresh ->
            {
                val provider = Provider(activity.contentResolver.query(providerUri, null, null, null, null))
                FetcherAsync.fetchProvider(provider, this, ::providerViewerComplete, ::providerViewerStatus, behavior = FetcherSync.Behavior.ForceRefresh)

                loading!!.progress = 0
                loading!!.show()
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onCreateLoader(id: Int, args: Bundle): Loader<Cursor>
    {
        //TODO:: Blechhh
        val seriesUri = providerUri!!.buildUpon().appendPath("series").build()
        if (filter == null)
        {
            return CursorLoader(activity, seriesUri, null, null, null, null)
        }
        else
        {
            return CursorLoader(activity, seriesUri, null, Series.nameCol + " like ?", arrayOf("%$filter%"), null)
        }
    }

    override fun onLoadFinished(loader: Loader<Cursor>, data: Cursor)
    {
        adapter!!.changeCursor(data)
    }

    override fun onLoaderReset(loader: Loader<Cursor>)
    {
        adapter!!.changeCursor(null)
    }

    override fun onListItemClick(l: ListView?, v: View?, position: Int, id: Long)
    {
        super.onListItemClick(l, v, position, id)
        parent!!.loadSeries(Series.uri(id.toInt()))
        val imm: InputMethodManager  = activity.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm.hideSoftInputFromWindow(v!!.windowToken, 0)
    }

    override fun onQueryTextSubmit(query: String): Boolean
    {
        return true
    }

    override fun onQueryTextChange(newText: String): Boolean
    {
        filter = if (!TextUtils.isEmpty(newText)) newText else null
        loaderManager.restartLoader(0, Bundle(), this)
        return true
    }

    companion object
    {
        private val PROVIDER_ARG_KEY: String = "ProviderUriArg"

        fun newInstance(uri: Uri): ProviderViewer
        {
            val providerViewer = ProviderViewer()
            val bundle = Bundle()
            bundle.putString(PROVIDER_ARG_KEY, uri.toString())
            providerViewer.arguments = bundle
            return providerViewer
        }
    }
}
